src/main/java/com/aws/greengrass/clientdevices/auth/certificate/infra/BackgroundCertificateRefresh.java

```diff
@@ -104,18 +129,144 @@ public void run() {
+        ThingAssociations associations = getThingsAssociatedWithCoreDevice();
+        refresh(associations);
+    }
+
+    /**
+     * Returns ThingAssociations which has information about what Things are associated with the core device and
+     * which things are no longer associated with the core device.
+     */
+    @SuppressWarnings("PMD.AvoidCatchingGenericException")
+    private ThingAssociations getThingsAssociatedWithCoreDevice() {
+        ThingAssociations associations = new ThingAssociations();
+        RetryUtils.RetryConfig retryConfig = RetryUtils.RetryConfig.builder()
+                .initialRetryInterval(Duration.ofSeconds(30)).maxRetryInterval(Duration.ofMinutes(3))
+                .maxAttempt(3)
+                .retryableExceptions(Arrays.asList(ThrottlingException.class, InternalServerException.class)).build();
+
+        try {
+            Set<String> cloudAttachedThingNames =  RetryUtils.runWithRetry(
+                    retryConfig, iotAuthClient::getThingsAssociatedWithCoreDevice,
+                            "get-things-associated-with-core-device", logger)
+                    .map(Thing::getThingName).collect(Collectors.toSet());
+
+            thingRegistry.getAllThings().forEach(thing -> {
+                if (cloudAttachedThingNames.contains(thing.getThingName()))  {
+                    associations.attached(thing);
+                } else {
+                    associations.detached(thing);
+                }
+            });
+        } catch (AccessDeniedException e) {
+            logger.atWarn().log(
+                    "Access denied to list things associated with core device. Please make sure you "
+                    + "have setup the correct permissions.");
+        } catch (Exception e) {
+            logger.warn("Failed to get things associated to the core device. Retry will be schedule later");
+        }
+
+        return associations;
+    }
 
-        for (String certificateId : certificateIds) {
-            if (pemStore.exists(certificateId)) {
-                String pem = pemStore.getPem(certificateId).get();
-                verifyIotCertificate.apply(pem);
+    private void refresh(ThingAssociations associations) {
+       associations.getThingAssociations().forEach(this::refreshCertificateThingAttachment);
+       associations.getCertificateAssociations().forEach(this::refreshCertificateValidity);
+       associations.getStaleCertificateAssociations().forEach(certificateRegistry::deleteCertificate);
+       associations.getStaleThingAssociations().forEach(thingRegistry::deleteThing);
+    }
+
+    private void refreshCertificateThingAttachment(Thing thing) {
+        VerifyThingAttachedToCertificate useCase = useCases.get(VerifyThingAttachedToCertificate.class);
+
+        thing.getAttachedCertificateIds().keySet().forEach(certificateId -> {
+            Optional<String> certPem = pemStore.getPem(certificateId);
+
+            if (!certPem.isPresent()) {
+                return;
```

Comment:
any thoughts on having a log here (may be TRACE)? It'd be useful to know why a certain certificate attachment was not refreshed. At this point if we don't have PEM for it, we may never have it until the client reconnects and until that time this cert won't be refreshed.